tskv/src/wal.rs

```diff
@@ -105,121 +354,235 @@ impl WriteAheadLogManager {
                 .split_at(1)
                 .1
                 .parse::<u64>()
-                .map_err(|err| Error::InvalidFileName {
+                .map_err(|err| WalError::InvalidFileName {
                     file_name: file_name.clone(),
                 })
         }
     }
 
-    fn get_wal_file<P: AsRef<Path>>(&self, path: P) -> Result<File> {
+    fn get_or_create_wal_file<P: AsRef<Path>>(
+        id: u64,
+        file_manager: &FileManager,
+        path: P,
+    ) -> WalResult<WalFile> {
+        let file = if file_manager::try_exists(path.as_ref()) {
+            file_manager.open_file(path)
+        } else {
+            file_manager.create_file(path)
+        };
+        let file = file
+            .and_then(|f| Ok(f))
+            .context(FailedWithFileManagerSnafu)?;
+
+        let size = file.len();
+
+        Ok(WalFile {
+            id,
+            file: Arc::new(file),
+            size,
+        })
+    }
+
+    fn get_file<P: AsRef<Path>>(&self, path: P) -> WalResult<File> {
         self.file_manager
             .open_file(path)
-            .map_err(|err| Error::FileManagerError { source: err })
+            .context(FailedWithFileManagerSnafu)
     }
 
-    fn new_wal_file(&mut self) -> Result<()> {
-        self.current_file_id = self.current_file_id + 1;
-        if self.current_file_writer.is_none() {
+    fn roll_wal_file(&mut self) -> WalResult<()> {
+        // let mut current_file = self.current_file.lock().unwrap();
+        let current_file = &mut self.current_file;
+        if current_file.size > SEGMENT_SIZE {
+            current_file.id += 1;
             let file_name = self
                 .current_dir_path
-                .join(format!("_{:05}.wal", self.current_file_id));
-            self.current_file_writer = Some(
-                self.file_manager
-                    .create_file(file_name)
-                    .map_err(|err| Error::FileManagerError { source: err })
-                    .map(|file| WriteAheadLogDiskFileWriter::new(file))?,
-            );
+                .join(format!("_{:05}.wal", current_file.id));
+            let file = self
+                .file_manager
+                .create_file(file_name)
+                .context(FailedWithFileManagerSnafu)?;
+            current_file.file = Arc::new(file);
         }
         Ok(())
     }
 
-    fn roll_wal_file(&mut self) -> Result<()> {
-        let writer = &self.current_file_writer;
-        match writer {
-            Some(w) if w.size > SEGMENT_SIZE => self.new_wal_file(),
-            None => self.new_wal_file(),
-            _ => Ok(()),
-        }
-    }
+    pub async fn write(&mut self, rows: &[u8]) -> WalResult<()> {
```

Comment:
restruct these three functions

Reply:
fixed in [09cf1f0](https://github.com/cnosdb/cnosdb/pull/237/commits/09cf1f08d3319e364443f8f512ac9a1ccdc77550)